include/deal.II/fe/fe_values_coupling.h

```diff
@@ -0,0 +1,557 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_fe_values_coupling_h
+#define dealii_fe_values_coupling_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/algorithms/general_data_storage.h>
+
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/fe/fe_values_views.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declaration
+#ifndef DOXYGEN
+template <int dim, int spacedim>
+class FEValuesBase;
+#endif
+
+namespace FEValuesViews
+{
+  /**
+   * A class that provides a renumbered view to a given FEValuesViews object.
+   *
+   * In general, the order of the degrees of freedom and quadrature points
+   * follows the one of the FEValues object. However, in some cases, it is
+   * convenient to group together degrees of freedom and quadrature points in a
+   * different order. This class provides a view to a given FEValuesViews
+   * object, where the degrees of freedom and quadrature points are renumbered
+   * according to the given renumbering vectors.
+   */
+  template <typename ViewType>
+  class ReorderedView : public Subscriptor
+  {
+  public:
+    /**
+     * An alias for the data type of values of the underlying view.
+     */
+    using value_type = typename ViewType::value_type;
+
+    /**
+     * An alias for the data type of gradients of the underlying view.
+     */
+    using gradient_type = typename ViewType::gradient_type;
+
+    /**
+     * An alias for the data type of the product of a @p Number and the values
+     * of the underlying view type.
+     */
+    template <typename Number>
+    using solution_value_type = typename ProductType<Number, value_type>::type;
+
+    /**
+     * An alias for the data type of the product of a @p Number and the gradients
+     * of the underlying view type.
+     */
+    template <typename Number>
+    using solution_gradient_type =
+      typename ProductType<Number, gradient_type>::type;
+
+    /**
+     * Construct a new ReorderedView object.
+     *
+     * The @p dof_renumbering vector is used to renumber the degrees of freedom,
+     * while the @p quadrature_renumbering vector is used to renumber the
+     * quadrature points. An empty renumbering vector simply means that no
+     * renumbering is performed.
+     *
+     * @note The renumbering vectors may *not* match the dimensions of the
```

Comment:
"may not" is often used to mean "not allowed to". Perhaps use the following?
```suggestion
     * @note The renumbering vectors are *not* required to match the dimensions of the
```